wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/ToolConverter.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server.convert.tool;
+
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.devfile.server.DevfileException;
+import org.eclipse.che.api.devfile.server.DevfileRecipeFormatException;
+import org.eclipse.che.api.devfile.server.FileContentProvider;
+import org.eclipse.che.api.devfile.server.WorkspaceExportException;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+import org.eclipse.che.commons.annotation.Nullable;
+
+/**
+ * Helps to inject/extract {@link Tool devfile tools} to/from {@link WorkspaceConfigImpl}.
+ *
+ * @author Sergii Leshchenko
+ */
+public interface ToolConverter {
+
+  /**
+   * Helps to inject {@link Tool devfile tools} to {@link WorkspaceConfigImpl}.
+   *
+   * <p>Note that implementation may change provided devfile if needed.
+   *
+   * @param workspaceConfig workspace config where tool should be injected
+   * @param devfile devfile that contains tools
+   * @param contentProvider optional file content provider that can be used for external tool
+   *     configuration retrieving.
+   * @throws IllegalArgumentException when there is no content provider for recipe-type tool
+   * @throws DevfileRecipeFormatException when recipe-type tool content is empty or has wrong format
+   * @throws DevfileException when general devfile error occurs
+   */
+  void toWorkspaceConfig(
```

Comment:
@garagatyi Do you have an idea what how it can be refactored?

Reply:
Yes, here are a couple of ideas:
- move filtering of tool out of the convert method to a separate method dedicated to checking if a tool needs to be converted by this converter. Maybe expose type of the converter as converter contract and let some class decide which converter should convert a particular tool/environment. 
- from config: make converter return tool instead of directly provisioning it in a devfile
- make sure converter doesn't modify entity which it converts (I don't say it does now, just saying it should not have side effects)
- separate conversion of machines, servers, volumes, commands into a separate methods to clean code

They might not be enough to make code really clean, but it is what I see as an improvement at the moment

Reply:
I believe I have done most of your suggestions in the latest commit.